### PR TITLE
Selective State Update kernel (mamba)

### PR DESCRIPTION
**Pull request by ishovkun**
## 📌 Description

This PR implements `selective_state_update` kernels for the Mamba layer (e.g. in the Nemotron model). This implementation is necessary as the default Triton implementation in TRT-LLM is inefficient. 

Specifically, the PR implements two kernels: 
- a simple pre-Hopper kernel with vectorized loads
- a producer-consumer kernel for Hopper+ GPUs.

The speedup yield of the CUDA code is summarized in the image below.
Screenshot 2026-01-06 at 3 05 13 PM

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Summary by CodeRabbit

* **New Features**
  * Public Mamba selective-state-update API: JIT-built CUDA backend with BF16 and SM90 variants, runtime backend selection/caching, optional dt-bias and z-scaling, state-batch indexing, dt-softplus, pad-slot handling, and multi-head batching with optimized device kernels.

* **Tests**
  * End-to-end pytest suite validating correctness against a Triton reference across dtypes, batching, and devices.

* **Chores**
  * Mamba package exposed at top level and integrated into AOT/JIT build flow.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: csrc/selective_state_update.cu [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+    TVM_FFI_ICHECK(false) << "Unsupported dtype combination for selective_state_update: "
+                          << "state_dtype=" << state_dtype.code << ":" << state_dtype.bits << ", "
+                          << "input_dtype=" << input_dtype.code << ":" << input_dtype.bits << ", "
+                          << "weight_dtype=" << weight_dtype.code << ":" << weight_dtype.bits
+                          << ", "
+                          << "matrixA_dtype=" << matrixA_dtype.code << ":" << matrixA_dtype.bits
+                          << ". Currently only support: "
+                          << "state=bfloat16, input=bfloat16, weight=bfloat16, matrixA=float32";
+
+    throw std::runtime_error("selective_state_update is not implemented yet.");
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply csrc/selective_state_update.cu]: Fixed in [1457e3f](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/7bb48ccaa1aeff09a8f5e15e04254608c1457e3f).

File: include/flashinfer/mamba/selective_state_update.cuh [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+        float out_value = toFloat(d_value) * x_value * int(lane == 0);  // first lane has the value
+
+        for (int i = lane; i < DSTATE; i += warpSize) {
+          auto const state_value = (state_batch != params.pad_slot_id)
+                                       ? toFloat(sram.state[stage][dd * DSTATE + i])
+                                       : 0.f;
+          auto const B_value = toFloat(sram.B[i]);
+          auto const C_value = toFloat(sram.C[i]);
+
+          auto const dA = fast_exp(A_value * dt_value);
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/selective_state_update.cuh]: Simple kernel now also uses fast_exp [#2301](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/9883eab7e27cb61568c8f1e62b1fd519e717e4e7).

File: include/flashinfer/mamba/selective_state_update.cuh [hunk 3]
```diff
@@ -0,0 +1,500 @@
...
+    auto tensorState =
+        tma::createTensorMap<state_t>(params.state, B * nh * dim, DSTATE, rowsPerStage, DSTATE);
+    FLASHINFER_CHECK(params.dim % rowsPerStage == 0);
+
+    scan_func<<<grid, block, 0, stream>>>(params, tensorState);
+    cudaError_t err = cudaGetLastError();
+    if (err != cudaSuccess) {
+      printf("Kernel launch failed: %s\n", cudaGetErrorString(err));
+    }
+  }
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/selective_state_update.cuh]: Removed the check in [#2301](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/8aed46837ff955bf519f88440e02f682d5e89e5c)

File: include/flashinfer/mamba/selective_state_update.cuh [hunk 4]
```diff
@@ -0,0 +1,500 @@
...
+  return __logf(1.f + __expf(x));
+  // return log1pf(exp(x));
+}
+
+__device__ __forceinline__ float fast_exp(float x) {
+  constexpr float log2_E = 1.4426950408889634f;
+  float y;
+  asm("ex2.approx.f32 %0,%1;" : "=f"(y) : "f"(x * log2_E));
+  return y;
+}
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/selective_state_update.cuh]: `__expf` is enough indeed and has a negligible performance impact. Fixed in [#2301](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/5c65e532a91b4ad8464501affccbf172f8b44a99).

File: flashinfer/mamba/selective_state_update.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+@flashinfer_api
+def selective_state_update(
+    state: torch.Tensor,
+    x: torch.Tensor,
+    dt: torch.Tensor,
+    A: torch.Tensor,
+    B: torch.Tensor,
+    C: torch.Tensor,
+    D: torch.Tensor,
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply flashinfer/mamba/selective_state_update.py:57]: This was fixed in [#2301](https://github.com/flashinfer-ai/flashinfer/pull/2301/changes/681a5bac4963295b95a3e8bbb8fcd69ddaa01975).

File: flashinfer/mamba/selective_state_update.py [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    z: Optional[torch.Tensor] = None,
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply flashinfer/mamba/selective_state_update.py:58]: Fixed in [8c63876](https://github.com/flashinfer-ai/flashinfer/pull/2301/changes/84bdc07a05ed2ca883c3e75d6d1abc29c8c63876)

[the remark this answers is not in the document]
Comment (ishovkun) [reply flashinfer/mamba/selective_state_update.py:58]: The code now supports z and it is checked with a new test.

File: include/flashinfer/mamba/selective_state_update.cuh [hunk 7]
```diff
@@ -0,0 +1,488 @@
...
+  constexpr auto warpSize = 32;
+  constexpr auto numWarps = 1 + consumerWarps;
+
+  auto const batch = blockIdx.x;
+  auto const head = blockIdx.y;
+  auto const group = head / (nheads / ngroups);
+  auto lane = threadIdx.x % warpSize;
+  auto warp = threadIdx.y;
+
+  auto const state_batch = (state_batch_indices) ? state_batch_indices[batch] : batch;
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/selective_state_update.cuh:275]: I added this check, but rather in the C++ flashinfer_check portion of the code (commit [07d2dc71d](https://github.com/flashinfer-ai/flashinfer/pull/2301/changes/3e9735474ad7eacfe8f57ff2a8cdfcd07d2dc71d)).

File: tests/mamba/test_selective_state_update.py [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+import torch
+
+import flashinfer
+
+from .selective_state_update_triton import selective_state_update_triton
+
+
+def create_test_inputs(
+    batch_size, nheads, dim, dstate, ngroups, input_dtype, weight_dtype, matrixA_dtype, z_none=True
+):
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply tests/mamba/test_selective_state_update.py:20]: fixed.

File: include/flashinfer/mamba/selective_state_update.cuh [hunk 9]
```diff
@@ -0,0 +1,502 @@
...
+    dim3 grid(blocks_per_dim, params.batch, params.nheads);
+    selective_state_update_kernel_simple<input_t, weight_t, matrixA_t, state_t, DSTATE, numWarps>
+        <<<grid, block, 0, stream>>>(params);
+  } else {
+    constexpr auto numConsumers = 4;
+    constexpr auto numWarps = 1 + numConsumers;
+    constexpr auto numStages = 3;
+    constexpr auto rowsPerStage = 4 * numConsumers;
+    constexpr auto DIM = 64;
+    FLASHINFER_CHECK(params.dim == DIM);
```
[the remark this answers is not in the document]
Comment (ishovkun) [reply include/flashinfer/mamba/selective_state_update.cuh]: I extended the code to handle: DSTATE [64, 128], and DIM[64, 128, 256].
See commit [40099d7543](https://github.com/flashinfer-ai/flashinfer/pull/2301/changes/ce2e8682a04d68ed35cc0658ddd4af40099d7543).

File: flashinfer/jit/mamba/selective_state_update.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+"""
+
+from .. import env as jit_env
+from ...compilation_context import CompilationContext
+from ..core import JitSpec, gen_jit_spec
+
+
+def gen_selective_state_update_module() -> JitSpec:
+    nvcc_flags = [
+        "-DENABLE_BF16",
```
Comment (yzh119) [inline flashinfer/jit/mamba/selective_state_update.py]: https://github.com/flashinfer-ai/flashinfer/blob/2062decf4f5465974806a547450e3a150d68718c/flashinfer/jit/core.py#L421

It's part of default flags, might not be necessary here.

File: flashinfer/jit/mamba/selective_state_update.py [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            jit_env.FLASHINFER_CSRC_DIR / "selective_state_update.cu",
+            jit_env.FLASHINFER_CSRC_DIR / "flashinfer_mamba_binding.cu",
+        ],
+        extra_cuda_cflags=nvcc_flags,
+    )
+
+
+def gen_selective_state_update_sm90_module() -> JitSpec:
+    # Use CompilationContext to get nvcc flags filtered by supported major versions
+    # This supports SM90 (Hopper) and future architectures
```
Comment (yzh119) [inline flashinfer/jit/mamba/selective_state_update.py:35]: Is it because of the use of tma?

File: flashinfer/jit/mamba/selective_state_update.py [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+    compilation_context = CompilationContext()
+    nvcc_flags = compilation_context.get_nvcc_flags_list(
+        supported_major_versions=[9, 10, 11, 12]
+    )
+    nvcc_flags += [
+        "-DENABLE_BF16",
```
Comment (yzh119) [inline flashinfer/jit/mamba/selective_state_update.py]: not necessary

File: tests/mamba/test_selective_state_update.py [hunk 13]
```diff
@@ -0,0 +1,309 @@
...
+        ssm_state_cache_size, nheads, dim, dstate, dtype=input_dtype, device=device
+    )
+
+    x = torch.randn(batch_size, nheads, dim, dtype=input_dtype, device=device)
+
+    dt = torch.randn(batch_size, nheads, dtype=weight_dtype, device=device).as_strided(
+        (batch_size, nheads, dim), (nheads, 1, 0)
+    )
+
+    A_base = -torch.rand(nheads, dtype=matrixA_dtype, device=device)
```
Comment (yzh119) [inline tests/mamba/test_selective_state_update.py]: Why do we use this distribution? Can you provide more insights?

File: tests/mamba/test_selective_state_update.py [hunk 14]
```diff
@@ -0,0 +1,309 @@
...
+
+    if outputs_match:
+        print(f"✓ Outputs match within tolerance (atol={atol}, rtol={rtol})")
+    else:
+        print(f"✗ Outputs do NOT match within tolerance (atol={atol}, rtol={rtol})")
+
+    assert outputs_match
+
+    # Check if states match within tolerance
+    states_match = torch.allclose(
```
Comment (yzh119) [inline tests/mamba/test_selective_state_update.py]: Can we use `torch.testing.assert_allclose` instead? It can display the number of mismatch elements and the location of mismatch elements.

File: tests/mamba/test_selective_state_update.py [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+            diff = abs(ref_val - test_val)
+            rel_diff = diff / (abs(ref_val) + 1e-8)
+            print(
+                f"  Index {idx_tuple}: ref={ref_val:.6f}, test={test_val:.6f}, diff={diff:.6e}, rel_diff={rel_diff:.6e}"
+            )
+
+    assert outputs_match
+
+    # Check if states match within tolerance
+    states_match = torch.allclose(
```
Comment (yzh119) [inline tests/mamba/test_selective_state_update.py:188]: ditto

File: flashinfer/jit/mamba/selective_state_update.py (hunk 10 above)
Comment (ishovkun) [reply flashinfer/jit/mamba/selective_state_update.py]: addressed in [2f5a774](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/dfc0a38c538a85e89d96674990014e2272f5a774)

File: flashinfer/jit/mamba/selective_state_update.py (hunk 11 above)
Comment (ishovkun) [reply flashinfer/jit/mamba/selective_state_update.py:35]: Yes. A corresponding comment has been added in [72f5a774](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/dfc0a38c538a85e89d96674990014e2272f5a774)

File: flashinfer/jit/mamba/selective_state_update.py (hunk 12 above)
Comment (ishovkun) [reply flashinfer/jit/mamba/selective_state_update.py]: addressed in [2f5a774](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/dfc0a38c538a85e89d96674990014e2272f5a774)

File: tests/mamba/test_selective_state_update.py (hunk 13 above)
Comment (ishovkun) [reply tests/mamba/test_selective_state_update.py]: The (-) minus was unnecessary. Added a comment about the shape and dtype of A in [011ef1ac4](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/5afc7f6527916c2d0ec83c61558a877011ef1ac4). 
TLDR: 
- The dtype of A is separate in nemotron nano v3. 
- A only has one value per head (as discussed in mamba 2 block) hence the strides.

File: tests/mamba/test_selective_state_update.py (hunk 14 above)
Comment (ishovkun) [reply tests/mamba/test_selective_state_update.py]: Done in [020f69](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/0d9c71b9f54479a6dc4a717a83eb166c83020f69).

File: tests/mamba/test_selective_state_update.py (hunk 15 above)
Comment (ishovkun) [reply tests/mamba/test_selective_state_update.py:188]: Done in [020f69](https://github.com/flashinfer-ai/flashinfer/pull/2301/commits/0d9c71b9f54479a6dc4a717a83eb166c83020f69).

**Comment by ishovkun** [comment]
> LGTM, thanks for this contribution!

I'll have a follow up soon with more perf for Blackwell!

**Comment by yzh119** [comment]
@ishovkun sounds great! Do you expect to land it in this PR, or in a follow up one?
